packages/blitz-dom/src/traversal.rs

```diff
@@ -235,4 +237,223 @@ impl BaseDocument {
+    /// Returns Ordering::Greater if node_a comes after node_b.
+    /// Returns Ordering::Equal if they are the same node.
+    pub fn compare_document_order(&self, node_a: usize, node_b: usize) -> Ordering {
+        if node_a == node_b {
+            return Ordering::Equal;
+        }
+
+        // Build ancestor chains from root to node (inclusive)
+        let chain_a = self.ancestor_chain_from_root(node_a);
+        let chain_b = self.ancestor_chain_from_root(node_b);
+
+        // Find where the chains diverge
+        let mut common_depth = 0;
+        for (a, b) in chain_a.iter().zip(chain_b.iter()) {
+            if a != b {
+                break;
+            }
+            common_depth += 1;
+        }
+
+        // If one is an ancestor of the other
+        if common_depth == chain_a.len() {
+            return Ordering::Less; // node_a is ancestor of node_b
+        }
+        if common_depth == chain_b.len() {
+            return Ordering::Greater; // node_b is ancestor of node_a
+        }
+
+        // Safety: common_depth must be > 0 here because both chains start from the same
+        // root node (node 0), so they share at least that node. If common_depth were 0,
+        // chain_a[0] != chain_b[0], but both start from root, so this is impossible.
+        debug_assert!(
+            common_depth > 0,
+            "nodes must share a common ancestor (the root)"
+        );
+
+        // Compare position among siblings at the divergence point
+        let divergent_a = chain_a[common_depth];
+        let divergent_b = chain_b[common_depth];
+        let parent_id = chain_a[common_depth - 1];
+        let parent = &self.nodes[parent_id];
+
+        for &child_id in &parent.children {
+            if child_id == divergent_a {
+                return Ordering::Less;
+            }
+            if child_id == divergent_b {
+                return Ordering::Greater;
+            }
+        }
+
+        // Should not reach here if tree is well-formed
+        Ordering::Equal
+    }
+
+    /// Build ancestor chain from root to node (inclusive), ordered [root, ..., node].
+    fn ancestor_chain_from_root(&self, node_id: usize) -> Vec<usize> {
+        let mut ancestors = Vec::with_capacity(16);
+        let mut current = Some(node_id);
+        while let Some(id) = current {
+            ancestors.push(id);
+            current = self.nodes[id].parent;
+        }
+        ancestors.reverse();
+        ancestors
+    }
+
+    /// Collect all inline root nodes between start_node and end_node in document order.
+    /// Both start and end are assumed to be inline roots.
+    /// Returns the nodes in document order (from first to last).
+    pub fn collect_inline_roots_in_range(&self, start_node: usize, end_node: usize) -> Vec<usize> {
+        let start = &self.nodes[start_node];
+        let end = &self.nodes[end_node];
+        let start_is_anon = start.is_anonymous();
+        let end_is_anon = end.is_anonymous();
```

Comment:
For this function, would it make sense to just resolve the current node id of the anonymous node? Then the following code wouldn't have to treat anonymous nodes specially?